File: Leetcode_solutions/n-ary-tree-preorder-traversal.js
/**
 * // Definition for a Node.
 * function Node(val,children) {
 *    this.val = val;
 *    this.children = children;
 * };
 */
/**
 * @param {Node} root
 * @return {number[]}
 */
var preorder = function(root) {
  if (root === null) {
    return [];
  }
  output = [[root.val]];
  search(root.children, output);
  return output;
};
function search(r, o) {
  if (r === null) {
    return o;
  }
  for (let i = 0; i < r.length; i++) {
    o.push(r[i].val);
    search(r[i].children, o);
  }
}

File: atcoder/N-ary Tree Preorder Traversal.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, children):
        self.val = val
        self.children = children
"""


class Solution:
    def preorder(self, root: 'Node') -> List[int]:
        ans = []
        self.search(root, ans)
        return ans

    def search(self, node, output):
        if node == None:
            return
        output.append(node.val)
        for i in range(len(node.children)):
            self.search(node.children[i], output)


"""
# Definition for a Node.
class Node:
    def __init__(self, val, children):
        self.val = val
        self.children = children
"""


class Solution(object):
    def preorder(self, root):
        """
        :type root: Node
        :rtype: List[int]
        """
        if root is None:
            return []

        stack, output = [root, ], []
        while stack:
            root = stack.pop()
            output.append(root.val)
            stack.extend(root.children[::-1])

        return output
